packages/amplify_datastore/ios/Classes/FlutterApiPlugin.swift

```diff
@@ -23,15 +34,96 @@ public class FlutterApiPlugin: APICategoryPlugin
         preconditionFailure("method not supported")
     }
     
-    // TODO: Implment in Async Swift v2
-    public func subscribe<R>(request: GraphQLRequest<R>) -> AmplifyAsyncThrowingSequence<GraphQLSubscriptionEvent<R>> where R : Decodable {
-        preconditionFailure("method not supported")
+    public func subscribe<R: Decodable>(
+        request: GraphQLRequest<R>
+    ) -> AmplifyAsyncThrowingSequence<GraphQLSubscriptionEvent<R>> where R : Decodable {
+        var subscriptionId: String? = ""
+        
+        // TODO: write a e2e test to ensure we don't go over 100 AppSync connections
+        func unsubscribe(subscriptionId: String?){
+            if let subscriptionId {
+                DispatchQueue.main.async {
+                    self.nativeApiPlugin.unsubscribe(subscriptionId: subscriptionId) {}
+                }
+            }
+        }
+        
+        // TODO: shouldn't there be a timeout if there is no start_ack returned in a certain period of time
+        let (sequence, cancellable) = nativeSubscriptionEvents
+            .filter { $0.subscriptionId == subscriptionId }
+            .handleEvents(receiveCompletion: {_ in
+                unsubscribe(subscriptionId: subscriptionId)
+            }, receiveCancel: {
+                unsubscribe(subscriptionId: subscriptionId)
+            })
+            .compactMap { [weak self] event -> GraphQLSubscriptionEvent<R>? in
+                switch event.type {
+                    case "connecting":
+                        return .connection(.connecting)
+                    case "start_ack":
+                        return .connection(.connected)
+                    case "complete":
+                        return .connection(.disconnected)
+                    case "data":
+                        if let responseDecoded: GraphQLResponse<R> =
+                            try? self?.decodeGraphQLPayloadJson(request: request, payload: event.payloadJson)
+                        {
+                            return .data(responseDecoded)
+                        }
+                        return nil
+                    case "error":
+                        if let payload = event.payloadJson {
+                            return .data(.fromAppSyncSubscriptionErrorResponse(string: payload))
+                        }
+                    return nil
+                    default:
+                        print("ERROR unsupported subscription event type! \(String(describing: event.type))")
+                        return nil
+                }
+            }
+            .toAmplifyAsyncThrowingSequence()
+        cancellables.insert(cancellable) // the subscription is bind with class instance lifecycle, it should be released when stream is finished or unsubscribed
+        sequence.send(.connection(.connecting))
+        DispatchQueue.main.async {
```

Comment:
do we really need to use `DispatchQueue.main` to perform the async task?

Reply:
Discussed offline. Flutter platform channel calls will be on main thread, and other logic will not. Adjusted logic accordingly.